### PR TITLE
Add uint1 to uint7 dtypes

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ pytorch/pytorch#117208

Summary:
These dtypes are added since we see more demand for these sub byte dtypes, especially with
the popularity of LLMs (https://pytorch.org/blog/accelerating-generative-ai-2/#step-4-reducing-the-size-of-the-weights-even-more-with-int4-quantization-and-gptq-2021-toks)

Note these are just placeholders, the operator support for these dtypes will be implemented with tensor subclass.
e.g. torch.empty(..., dtype=torch.uint1) will return a tensor subclass of uint1, that supports different operations like bitwsise ops, add, mul etc. (will be added later)

Also Note that these are not quantized data types, we'll implement quantization logic with tensor subclass backed up by these dtypes as well.
e.g `Int4GroupedQuantization(torch.Tensor)` will be implemented with torch.uint4 Tensors (see https://github.com/pytorch-labs/ao/pull/13 as an example)

Test Plan:
CIs
python test/test_quantization.py -k test_uint1_7_dtype 

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: c10/core/ScalarType.h [hunk 1]
```diff
@@ -77,7 +78,14 @@ namespace c10 {
   _(c10::Float8_e4m3fnuz, Float8_e4m3fnuz) /* 26 */      \
   _(uint16_t, UInt16) /* 27 */                           \
   _(uint32_t, UInt32) /* 28 */                           \
-  _(uint64_t, UInt64) /* 29 */
+  _(uint64_t, UInt64) /* 29 */                           \
+  _(c10::uint1_t, UInt1) /* 30 */                        \
```
Comment (ezyang) [inline c10/core/ScalarType.h:82]: I would prefer these to have dummy in the name to indicate that they don't actually work. It should also be feasible to define only one struct templates over int

**Review by ezyang — COMMENTED** [review]
Add a test that the Python bindings are working and you can make a wrapper subclass with it

File: aten/src/ATen/Dispatch_v2.h [hunk 2]
```diff
@@ -112,7 +112,7 @@
 
 // Ensure we never have too many scalar types for the expansion here to
 // support.  To bump this, you must regenerate the macros below.
-static_assert(static_cast<int>(c10::ScalarType::NumOptions) < 32);
+static_assert(static_cast<int>(c10::ScalarType::NumOptions) < 45);
```
Comment (albanD) [inline aten/src/ATen/Dispatch_v2.h:115]: nit: change num_args in the python code below to 45

**Comment by vadimkantorov** [comment]
Should there also be opaque dtypes for 2-bit/3-bit/4-bit sub-bytes? I think, the quantization efforts now go even there... Regarding 1-bit, there also exists "One-bit Adam" (I think even implemented somewhere in Meta experimental optimizer repos) which use the 1-bit

One nasty thing is that if you google uint1 or uint4, 1-byte and 4-byte exotic namings come out: https://people.montefiore.uliege.be/boigelot/research/lash/man/uint.html which is not very nice...

**Comment by jerryzh168** [comment]
> Should there also be opaque dtypes for 2-bit/3-bit/4-bit sub-bytes?

yeah we have these in the PR, this PR adds all dtypes from uint1 to uint7

> One nasty thing is that if you google uint1 or uint4, 1-byte and 4-byte exotic namings come out: [people.montefiore.uliege.be/boigelot/research/lash/man/uint.html](https://people.montefiore.uliege.be/boigelot/research/lash/man/uint.html) which is not very nice...

do you mean the lash-dtype naming?


> What is the pre-supposed indexing semantics for these sub-byte types? (and if some standardized indexing is supposed at all?) e.g. uint1tensor[3] should retrieve the 3-rd bit of the first byte? or the third encompassing byte?

I don't think we want to support sub byte indexing or non-byte aligned indexing. but we can support byte aligned indexing by unpacking, see: https://github.com/pytorch-labs/ao/pull/13/files#diff-109a7f01577eb57b0d9facb5e1c17c23158f544b7203cda513075487a389b2f6R160-R165

**Comment by vadimkantorov** [comment]
Yeah, the indexing / virtual vs actual byte shape is important for the usecase like BitTensor/BitMap where semantically they want to be like compressed BoolTensor

**Comment by jerryzh168** [comment]
> Regarding the naming, I mean that uint1/uint4 probably would google badly, as currently there exist some trash mentions in other contexts where 1/4 stand for byte and not bit, and uint4_t does not exist in C context.
> 
I think uint1 to uint7 is consistent with C/C++ naming of uint8, uint16, uint32 dtypes

> Regarding the uint1, I would also suggest to have some alias or a subclass in core like torch.bit or torch.bitmap or torch.bitset or similar which suggest also a higher-level usage - the compressing BoolTensor might a relatively frequent high-level usecase, also with pack/unpack and RoaringBitmap-like ops, and then maybe some classical morphological / binary image processing ops, also maybe with some LSH/hashing ops

can you write some quick code example of what you want to do (what is higher-level usage)? I think these may be built in tensor subclass in general

**Comment by ezyang** [comment]
> Yeah, the indexing / virtual vs actual byte shape is important for the usecase like BitTensor/BitMap where semantically they want to be like compressed BoolTensor

For non-aligned indexing, I think the most plausible implementation strategy in this direction is an introduction of a 1/8 SymInt (similar to SingletonSymNode we use to represent ragged dimension). Then, if I have a 8 element uint1 tensor, the storage offset of the 1-index element is 1/8, 2-index is 2/8, and so forth.

TBH, @jerryzh168 and co are not that interested in the packed bool tensor use case, so someone else is probably going to have to implement it.

> Regarding the uint1, I would also suggest to have some alias or a subclass in core like torch.bit or torch.bitmap or torch.bitset or similar which suggest also a higher-level usage - the compressing BoolTensor might a relatively frequent high-level usecase, also with pack/unpack and RoaringBitmap-like ops, and then maybe some classical morphological / binary image processing ops, also maybe with some LSH/hashing ops

My vote is for torch.bit

File: test/quantization/core/test_utils.py [hunk 3]
```diff
@@ -191,3 +191,30 @@ def test_quantize_weight_clamping_per_channel(self):
...
+            @classmethod
+            def __torch_dispatch__(cls, func, types, args, kwargs=None):
+                pass
+
+        # make sure it runs
+        x = UInt4Tensor(torch.tensor([
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+        ], dtype=torch.uint8))
```
Comment (ezyang) [inline test/quantization/core/test_utils.py:220]: test that x.dtype == torch.uint4

File: torch/csrc/utils/tensor_dtypes.cpp [hunk 4]
```diff
@@ -12,6 +12,20 @@ namespace utils {
 
 std::pair<std::string, std::string> getDtypeNames(at::ScalarType scalarType) {
   switch (scalarType) {
+    case at::ScalarType::UInt1:
+      return std::make_pair("uint1", "");
```
Comment (ezyang) [inline torch/csrc/utils/tensor_dtypes.cpp:16]: To accept the "bit" suggestion, change this to `"uint1", "bit"`

**Review by ezyang — APPROVED** [review]

File: torch/csrc/utils/tensor_dtypes.cpp (hunk 4 above)
Comment (jerryzh168) [reply torch/csrc/utils/tensor_dtypes.cpp:16]: OK. not sure how this works, do I use bit for all uint1 to uint7? and what does that mean?

Comment (jerryzh168) [reply torch/csrc/utils/tensor_dtypes.cpp:16]: this looks like just an alias? do you mean to have bit1 to bit7?

Comment (vadimkantorov) [reply torch/csrc/utils/tensor_dtypes.cpp:16]: I think, it's to introduce a more discoverable alias for `uint1` = `bit1` :)

**Comment by vadimkantorov** [comment]
> what is higher-level usage

e.g. as pack/unpack for compressing BoolTensor, set ops (RoaringBitmaps style), binary image processing, Hadamard matrices, etc :)

File: torch/csrc/utils/tensor_dtypes.cpp (hunk 4 above)
Comment (ezyang) [inline torch/csrc/utils/tensor_dtypes.cpp:16]:
```suggestion
      return std::make_pair("uint1", "bit");
```

Only this

Comment (jerryzh168) [reply torch/csrc/utils/tensor_dtypes.cpp:16]: oh OK

**Pull request closed without merging**